### PR TITLE
Add Support For Authentication Methods endpoints 

### PR DESCRIPTION
**Pull request by MarkDordoy**
This PR adds support for the authentication methods endpoint found here: https://docs.microsoft.com/en-us/graph/api/resources/authenticationmethods-overview?view=graph-rest-beta

### CODE REVIEW

File: msgraph/authentication_methods.go [hunk 1]
```diff
@@ -0,0 +1,815 @@
...
+}
+
+// NewAuthenticationClient returns a new ApplicationsClient
+func NewAuthenticationMethodsClient(tenantId string) *AuthenticationMethodsClient {
+	return &AuthenticationMethodsClient{
+		BaseClient: NewClient(VersionBeta, tenantId),
+	}
+}
+
+//List all authentication methods
```
Comment (manicminer) [inline msgraph/authentication_methods.go:26]: The linter should have noted the comment alignment here :)

File: msgraph/authentication_methods.go [hunk 2]
```diff
@@ -0,0 +1,815 @@
...
+		},
+	})
+	if err != nil {
+		return status, fmt.Errorf("AuthenticationMethodsClient.BaseClient.Delete(): %v", err)
+	}
+
+	return status, nil
+}
+
+func (c *AuthenticationMethodsClient) ListMSAuthenticatorMethods(ctx context.Context, userID string, query odata.Query) (*[]MicrosoftAuthenticatorAuthenticationMethod, int, error) {
```
Comment (manicminer) [inline msgraph/authentication_methods.go:195]: Could we name this `ListMicrosoftAuthenticatorMethods`? It's long, but more consistent

```suggestion
func (c *AuthenticationMethodsClient) ListMicrosoftAuthenticatorMethods(ctx context.Context, userID string, query odata.Query) (*[]MicrosoftAuthenticatorAuthenticationMethod, int, error) {
```

File: msgraph/authentication_methods.go [hunk 3]
```diff
@@ -0,0 +1,815 @@
...
+		MicrosoftAuthenticatorMethods []MicrosoftAuthenticatorAuthenticationMethod `json:"value"`
+	}
+	if err := json.Unmarshal(respBody, &data); err != nil {
+		return nil, status, fmt.Errorf("json.Unmarshal(): %v", err)
+	}
+
+	return &data.MicrosoftAuthenticatorMethods, status, nil
+}
+
+func (c *AuthenticationMethodsClient) GetMSAuthenticatorMethod(ctx context.Context, userID, id string, query odata.Query) (*MicrosoftAuthenticatorAuthenticationMethod, int, error) {
```
Comment (manicminer) [inline msgraph/authentication_methods.go:225]: Here too

```suggestion
func (c *AuthenticationMethodsClient) GetMicrosoftAuthenticatorMethod(ctx context.Context, userID, id string, query odata.Query) (*MicrosoftAuthenticatorAuthenticationMethod, int, error) {
```

File: msgraph/authentication_methods.go [hunk 4]
```diff
@@ -0,0 +1,815 @@
...
+
+	var microsoftAuthenticatorMethod MicrosoftAuthenticatorAuthenticationMethod
+	if err := json.Unmarshal(respBody, &microsoftAuthenticatorMethod); err != nil {
+		return nil, status, fmt.Errorf("json.Unmarshal(): %v", err)
+	}
+
+	return &microsoftAuthenticatorMethod, status, nil
+}
+
+func (c *AuthenticationMethodsClient) DeleteMSAuthenticatorMethod(ctx context.Context, userID, id string) (int, error) {
```
Comment (manicminer) [inline msgraph/authentication_methods.go:253]: Also here

```suggestion
func (c *AuthenticationMethodsClient) DeleteMicrosoftAuthenticatorMethod(ctx context.Context, userID, id string) (int, error) {
```

File: msgraph/authentication_methods.go [hunk 5]
```diff
@@ -0,0 +1,815 @@
...
+		},
+	})
+	if err != nil {
+		return status, fmt.Errorf("AuthenticationMethodsClient.BaseClient.Delete(): %v", err)
+	}
+
+	return status, nil
+}
+
+func (c *AuthenticationMethodsClient) ListTempAccessPassMethods(ctx context.Context, userID string, query odata.Query) (*[]TemporaryAccessPassAuthenticationMethod, int, error) {
```
Comment (manicminer) [inline msgraph/authentication_methods.go:343]: Similarly:

```suggestion
func (c *AuthenticationMethodsClient) ListTemporaryAccessPassMethods(ctx context.Context, userID string, query odata.Query) (*[]TemporaryAccessPassAuthenticationMethod, int, error) {
```

File: msgraph/authentication_methods.go [hunk 6]
```diff
@@ -0,0 +1,815 @@
...
+		TempAccessPassMethods []TemporaryAccessPassAuthenticationMethod `json:"value"`
+	}
+	if err := json.Unmarshal(respBody, &data); err != nil {
+		return nil, status, fmt.Errorf("json.Unmarshal(): %v", err)
+	}
+
+	return &data.TempAccessPassMethods, status, nil
+}
+
+func (c *AuthenticationMethodsClient) GetTempAccessPassMethod(ctx context.Context, userID, id string, query odata.Query) (*TemporaryAccessPassAuthenticationMethod, int, error) {
```
Comment (manicminer) [inline msgraph/authentication_methods.go:373]:
```suggestion
func (c *AuthenticationMethodsClient) GetTemporaryAccessPassMethod(ctx context.Context, userID, id string, query odata.Query) (*TemporaryAccessPassAuthenticationMethod, int, error) {
```

File: msgraph/authentication_methods.go [hunk 7]
```diff
@@ -0,0 +1,815 @@
...
+
+	var temporaryAccessPassMethod TemporaryAccessPassAuthenticationMethod
+	if err := json.Unmarshal(respBody, &temporaryAccessPassMethod); err != nil {
+		return nil, status, fmt.Errorf("json.Unmarshal(): %v", err)
+	}
+
+	return &temporaryAccessPassMethod, status, nil
+}
+
+func (c *AuthenticationMethodsClient) CreateTempAccessPassMethod(ctx context.Context, userID string, accessPass TemporaryAccessPassAuthenticationMethod) (*TemporaryAccessPassAuthenticationMethod, int, error) {
```
Comment (manicminer) [inline msgraph/authentication_methods.go:401]:
```suggestion
func (c *AuthenticationMethodsClient) CreateTemporaryAccessPassMethod(ctx context.Context, userID string, accessPass TemporaryAccessPassAuthenticationMethod) (*TemporaryAccessPassAuthenticationMethod, int, error) {
```

File: msgraph/authentication_methods.go [hunk 8]
```diff
@@ -0,0 +1,815 @@
...
+
+	var newTempAccessPassAuthMethod TemporaryAccessPassAuthenticationMethod
+	if err := json.Unmarshal(respBody, &newTempAccessPassAuthMethod); err != nil {
+		return nil, status, fmt.Errorf("json.Unmarshal(): %v", err)
+	}
+
+	return &newTempAccessPassAuthMethod, status, nil
+}
+
+func (c *AuthenticationMethodsClient) DeleteTempAccessPassMethod(ctx context.Context, userID, id string) (int, error) {
```
Comment (manicminer) [inline msgraph/authentication_methods.go:435]:
```suggestion
func (c *AuthenticationMethodsClient) DeleteTemporaryAccessPassMethod(ctx context.Context, userID, id string) (int, error) {
```

File: msgraph/authentication_methods.go [hunk 9]
```diff
@@ -0,0 +1,815 @@
...
+		Body:                   body,
+		ConsistencyFailureFunc: RetryOn404ConsistencyFailureFunc,
+		ValidStatusCodes:       []int{http.StatusOK},
+		Uri: Uri{
+			Entity:      fmt.Sprintf("/users/%s/authentication/phoneMethods/%s", userID, *phone.ID),
+			HasTenantId: true,
+		},
+	})
+	if err != nil {
+		return status, fmt.Errorf("AuthenticationMethodsClient.BaseClient.Patch(): %v", err)
```
Comment (manicminer) [inline msgraph/authentication_methods.go:581]:
```suggestion
		return status, fmt.Errorf("AuthenticationMethodsClient.BaseClient.Put(): %v", err)
```

File: msgraph/authentication_methods.go [hunk 10]
```diff
@@ -0,0 +1,815 @@
...
+	_, status, _, err := c.BaseClient.Post(ctx, PostHttpRequestInput{
+		ConsistencyFailureFunc: RetryOn404ConsistencyFailureFunc,
+		ValidStatusCodes:       []int{http.StatusOK},
+		Uri: Uri{
+			Entity:      fmt.Sprintf("/users/%s/authentication/phoneMethods/%s/enableSmsSignIn", userID, ID),
+			HasTenantId: true,
+		},
+	})
+	if err != nil {
+		return status, fmt.Errorf("AuthenticationMethodsClient.BaseClient.Patch(): %v", err)
```
Comment (manicminer) [inline msgraph/authentication_methods.go:599]:
```suggestion
		return status, fmt.Errorf("AuthenticationMethodsClient.BaseClient.Post(): %v", err)
```

File: msgraph/authentication_methods.go [hunk 11]
```diff
@@ -0,0 +1,815 @@
...
+	_, status, _, err := c.BaseClient.Post(ctx, PostHttpRequestInput{
+		ConsistencyFailureFunc: RetryOn404ConsistencyFailureFunc,
+		ValidStatusCodes:       []int{http.StatusOK},
+		Uri: Uri{
+			Entity:      fmt.Sprintf("/users/%s/authentication/phoneMethods/%s/disableSmsSignIn", userID, ID),
+			HasTenantId: true,
+		},
+	})
+	if err != nil {
+		return status, fmt.Errorf("AuthenticationMethodsClient.BaseClient.Patch(): %v", err)
```
Comment (manicminer) [inline msgraph/authentication_methods.go:617]:
```suggestion
		return status, fmt.Errorf("AuthenticationMethodsClient.BaseClient.Post(): %v", err)
```

File: msgraph/authentication_methods.go [hunk 12]
```diff
@@ -0,0 +1,815 @@
...
+		Body:                   body,
+		ConsistencyFailureFunc: RetryOn404ConsistencyFailureFunc,
+		ValidStatusCodes:       []int{http.StatusOK},
+		Uri: Uri{
+			Entity:      fmt.Sprintf("/users/%s/authentication/emailMethods/%s", userID, *email.ID),
+			HasTenantId: true,
+		},
+	})
+	if err != nil {
+		return status, fmt.Errorf("AuthenticationMethodsClient.BaseClient.Patch(): %v", err)
```
Comment (manicminer) [inline msgraph/authentication_methods.go:703]:
```suggestion
		return status, fmt.Errorf("AuthenticationMethodsClient.BaseClient.Put(): %v", err)
```

**Review by manicminer — COMMENTED** [review]
Hi @MarkDordoy, this is looking great. Just a few minor comments on naming etc. Once we have tests for these, should be good to merge 👍

**Comment by MarkDordoy** [comment]
Thanks for reviewing @manicminer. I've added your suggested commits. I'll re-review this code now and then get some tests written.

File: msgraph/authentication_methods_test.go [hunk 13]
```diff
@@ -0,0 +1,391 @@
...
+	_ = testAuthMethods_ListMicrosoftAuthenticatorMethods(t, c, *user.ID)
+	_ = testAuthMethods_ListWindowsHelloMethods(t, c, *user.ID)
+	tempAccessPass := testAuthMethods_CreateTemporaryAccessPassMethod(t, c, *user.ID)
+	_ = testAuthMethods_GetTemporaryAccessPassMethod(t, c, *user.ID, *tempAccessPass.ID)
+	_ = testAuthMethods_ListTemporaryAccessPassMethods(t, c, *user.ID)
+	testAuthMethods_DeleteTemporaryAccessPassMethod(t, c, *user.ID, *tempAccessPass.ID)
+	phoneAuthMethod := testAuthMethods_CreatePhoneMethod(t, c, *user.ID)
+	_ = testAuthMethods_GetPhoneMethod(t, c, *user.ID, *phoneAuthMethod.ID)
+	_ = testAuthMethods_ListPhoneMethods(t, c, *user.ID)
+	phoneAuthMethod.PhoneNumber = utils.StringPtr("+44 07940123966")
```
Comment (manicminer) [inline msgraph/authentication_methods_test.go:60]: @MarkDordoy I'm guessing this is a fake number? Is it a real dummy number or could someone actually have this number somewhere?

File: msgraph/authentication_methods_test.go [hunk 14]
```diff
@@ -0,0 +1,391 @@
...
+	testAuthMethods_UpdatePhoneMethod(t, c, *user.ID, *phoneAuthMethod)
+	testAuthMethods_EnablePhoneSMS(t, c, *user.ID, *phoneAuthMethod.ID)
+	testAuthMethods_DisablePhoneSMS(t, c, *user.ID, *phoneAuthMethod.ID)
+	testAuthMethods_DeletePhoneMethod(t, c, *user.ID, *phoneAuthMethod.ID)
+	emailAuthMethod := testAuthMethods_CreateEmailMethod(t, c, *user.ID)
+	_ = testAuthMethods_GetEmailMethod(t, c, *user.ID, *emailAuthMethod.ID)
+	_ = testAuthMethods_ListEmailMethods(t, c, *user.ID)
+	emailAuthMethod.EmailAddress = utils.StringPtr("test-user-authenticationmethods@updateddomain.com")
```
Comment (manicminer) [inline msgraph/authentication_methods_test.go:68]: @MarkDordoy Would prefer to use contoso.com or similar well-contained domain - other domains could feasibly be registered and potentially [ab]used

File: msgraph/authentication_methods_test.go (hunk 13 above)
Comment (MarkDordoy) [reply msgraph/authentication_methods_test.go:60]: Hey, yes i made it up.

Comment (manicminer) [reply msgraph/authentication_methods_test.go]: No worries, I changed it to a test number (reserved by Ofcom for media/testing etc)

File: msgraph/authentication_methods_test.go (hunk 14 above)
Comment (manicminer) [reply msgraph/authentication_methods_test.go]: Pushed :)

File: msgraph/authentication_methods_test.go [hunk 15]
```diff
@@ -0,0 +1,391 @@
...
+	_ = testAuthMethods_ListMicrosoftAuthenticatorMethods(t, c, *user.ID)
+	_ = testAuthMethods_ListWindowsHelloMethods(t, c, *user.ID)
+	tempAccessPass := testAuthMethods_CreateTemporaryAccessPassMethod(t, c, *user.ID)
+	_ = testAuthMethods_GetTemporaryAccessPassMethod(t, c, *user.ID, *tempAccessPass.ID)
+	_ = testAuthMethods_ListTemporaryAccessPassMethods(t, c, *user.ID)
+	testAuthMethods_DeleteTemporaryAccessPassMethod(t, c, *user.ID, *tempAccessPass.ID)
+	phoneAuthMethod := testAuthMethods_CreatePhoneMethod(t, c, *user.ID)
+	_ = testAuthMethods_GetPhoneMethod(t, c, *user.ID, *phoneAuthMethod.ID)
+	_ = testAuthMethods_ListPhoneMethods(t, c, *user.ID)
+	phoneAuthMethod.PhoneNumber = utils.StringPtr("+44 07700900000")
```
Comment (MarkDordoy) [inline msgraph/authentication_methods_test.go:60]: I ran that number through graph explorer in my tenant. I got the following error:
```
"message": "The provided phone number is formatted properly, but is an invalid phone number.",
```

Comment (manicminer) [reply msgraph/authentication_methods_test.go]: Yeah same here :(

Seems like the API knows about Ofcom reserved numbers. I've just made up an unlikely one for now.

**Review by manicminer — APPROVED** [review]
LGTM 🚀

**Pull request merged**